src/main/java/org/folio/bulkops/service/QueryService.java

```diff
@@ -1,94 +1,318 @@
+    bulkOperationRepository.save(bulkOperation);
+    return bulkOperation;
+  }
+
+  private UUID executeQuery(SubmitQuery submitQuery) {
     return queryClient.executeQuery(submitQuery).getQueryId();
   }
 
-  public BulkOperation checkQueryExecutionStatus(BulkOperation bulkOperation) {
-    var queryResult = queryClient.getQuery(bulkOperation.getFqlQueryId());
-    return switch (queryResult.getStatus()) {
-      case SUCCESS -> {
-        if (queryResult.getTotalRecords() == 0) {
-          yield failBulkOperation(bulkOperation, "No records found for the query");
-        }
-        executor.execute(getRunnableWithCurrentFolioContext(() -> saveIdentifiers(bulkOperation)));
-        bulkOperation.setStatus(RETRIEVING_IDENTIFIERS);
-        yield bulkOperationRepository.save(bulkOperation);
-      }
-      case FAILED -> failBulkOperation(bulkOperation, queryResult.getFailureReason());
-      case CANCELLED -> cancelBulkOperation((bulkOperation));
-      case IN_PROGRESS -> bulkOperation;
-    };
+  private QueryDetails getResult(UUID queryId) {
+    return queryClient.getQuery(queryId, true);
   }
 
-  private void saveIdentifiers(BulkOperation bulkOperation) {
+  private void saveIdentifiersAndStartBulkOperation(BulkOperation bulkOperation, QueryDetails queryResult) {
     try {
-      var identifiersString = queryClient.getSortedIds(bulkOperation.getFqlQueryId(), 0, Integer.MAX_VALUE).stream()
-        .map(list -> list.get(0))
-        .distinct()
-        .collect(Collectors.joining(NEW_LINE_SEPARATOR));
+      var identifiers = retrieveIdentifiers(queryResult, getIdField(bulkOperation));
+      var identifiersString = String.join(NEW_LINE_SEPARATOR, identifiers);
       var path = String.format(QUERY_FILENAME_TEMPLATE, bulkOperation.getId());
       remoteFileSystemClient.put(new ByteArrayInputStream(identifiersString.getBytes()), path);
       bulkOperation.setLinkToTriggeringCsvFile(path);
       bulkOperation.setStatus(SAVED_IDENTIFIERS);
+      bulkOperation.setTotalNumOfRecords(identifiers.size());
       bulkOperationRepository.save(bulkOperation);
+      startQueryOperation(queryResult, bulkOperation);
     } catch (Exception e) {
       var errorMessage = "Failed to save identifiers, reason: " + e.getMessage();
       log.error(errorMessage);
       failBulkOperation(bulkOperation, errorMessage);
     }
   }
 
+  private String getIdField(BulkOperation operation) {
+    return switch (operation.getEntityType()) {
+        case USER -> "users.id";
+        case ITEM -> "items.id";
+        case HOLDINGS_RECORD -> "holdings.id";
+        case INSTANCE, INSTANCE_MARC -> "instance.id";
+    };
+  }
+
+  private void startQueryOperation(QueryDetails queryResult, BulkOperation operation) {
+    try {
+      var matchedJsonFileName = operation.getId() + SLASH + "json" + SLASH + LocalDate.now() + MATCHED_RECORDS + FilenameUtils.getBaseName(operation.getLinkToTriggeringCsvFile()) + ".json";
```

Comment:
1) It will be nice to have utils methods for files names constructing; 2) Approach with file name patterns should be used to unify file names generating approach (for example, https://github.com/folio-org/mod-bulk-operations/blob/d35032577575936aceea581c1cea27f7afa31cea/src/main/java/org/folio/bulkops/service/BulkOperationService.java#L161)

Reply:
Done.